Fix UnboundLocalError remote_head in git

robinro opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
git

##### ANSIBLE VERSION
```devel```

##### SUMMARY

Fixes ansible/ansible-modules-core#5505

The use of remote_head was a leftover of ansible/ansible-modules-core#4562.
remote_head is not necessary, since the repo is unchanged anyway and
after is set correctly.

Further changes:
* Set changed=True and msg once local_mods are detected and reset.
* Remove need_fetch that is always True (due to previous if) to improve
clarity.
* Don't exit early for local_mods but run submodules update and
switch_version.

Commit:
Fix UnboundLocalError remote_head in git

Fixes #5505

The use of remote_head was a leftover of #4562.
remote_head is not necessary, since the repo is unchanged anyway and
after is set correctly.

Further changes:
* Set changed=True and msg once local_mods are detected and reset.
* Remove need_fetch that is always True (due to previous if) to improve
clarity
* Don't exit early for local_mods but run submodules update and
switch_version

robinro commented:
Migrated from https://github.com/ansible/ansible-modules-core/pull/5720

gundalow commented:
From the original PR
> Since this affects quite a few people I would suggest to also have @jurgenhaas or one of the other commenters make sure this covers their use case before merging it.

jurgenhaas commented:
I've tested this approach and can confirm that my particular use case is working with this one as well.

gundalow commented:
Thanks for the review. 
I wonder if this should go into `stable-2.2`, will discuss in Core Meeting on Tuesday (it's already on the agenda)

abadger commented:
@robinro I just started looking at migrating this: https://github.com/ansible/ansible/pull/19257  since @ryansb had given that one a +1.  (original PR: https://github.com/ansible/ansible-modules-core/pull/5260/files ).  Do these two approaches need to be merged?  Is part of https://github.com/ansible/ansible/pull/19257  needed but hte other part unrelated?

robinro commented:
From a quick look I can't think of a usecase that #19257 improves over this one, but I might be missing something. Maybe @sbussetti or @ryansb have more to add?

In general we should have integration tests for these codepaths.

sbussetti commented:
@robinro -- yeah looks like ships passing in the night.

I ran into this branch of code with the bad scoping for `remote_head` as a result of doing shallow forced checkouts.  Honestly I had a tough time figuring a way to write a test for that, (any examples of a similar test would be warmly welcomed), however, I'm gonna live test #19057 now and report back.

sbussetti commented:
So, this pull fixes the scoping issue, but there's a latent problem here that my pull also addressed.  I'm using a directive like the following to deploy my source:

```
    - name: Deploy Source
      become: true
      become_user: "{{deploy_user}}"
      git:
        accept_hostkey=True
        depth=1
        force=True
        repo="{{repo_url}}"
        dest="{{www_root}}"
        remote="{{repo_remote}}"
        version="{{repo_sha|default(repo_branch, true)}}"
        refspec="{{repo_ref|default('refs/heads/' ~ repo_branch, true)}}"
        key_file="/home/{{deploy_user}}/.ssh/id_rsa-deploy"
      tags:
        - source

```

And the git module skips one critical step (on the devel and on robinro's) branch -- that being actually fetching the requested branch / spec / version.   Since it never fetches, it just hard resets to whatever was available the first time it checked out the repo.  Subsequent runs get "stuck" on that earlier commit.

My adjustments to reset, here: https://github.com/ansible/ansible/pull/19257/files#diff-423dabf0f8d841053684543485e386c9L435

Which were related to this but also smoothed over the issue because without fetching the appropriate head, you're force resetting to the wrong thing.

It seems like I need to submit a new PR to just address that?  Please advise.

robinro commented:
For tests I would suggest to use git repos that are setup within the test role itself. So far we rely mostly on external repositories, which make it harder to understand how the testcase was constructed and adds a dependency on the network/github/some-private-repo being available.

An example I like is #18728. You write a shell one-liner that sets up your repository/... and then you run the git module to test it. One can also clone a repo and then run some shell commands to add local modifications. I'd be happy to review and assist with those tests.

abadger commented:
@robinro So it sounds like we want this change to fix remote_head and then part of https://github.com/ansible/ansible-modules-core/pull/5260 to fix the fetching/reset problem?

@sbussetti Feel free to migrate/redo https://github.com/ansible/ansible-modules-core/pull/5260 to do that and either you or robinro can ping me in irc to make sure I see it and merge it.  And yes, tets are very much appreciated.  The git module code has many cornercases that we run the risk of re-introducing without tests :-)

Commit:
Add test for git with local modifications

robinro commented:
I added a testcase for local modifications and overwriting them with force while fetching a new commit from a remote repo. Both for full-history and shallow clones. They both pass with the changes made here.

@sbussetti I don't see the "getting stuck on an old commit" situation in that test. Can you give a (minimal) working example that fails? Maybe you can construct a test based on the one I added here?

sbussetti commented:
@robinro  On it -- thank you for the guidance so far.

sbussetti commented:
Okay, 3d52ae53f17928cde7d9d14e101a83c0e33b0384 provides an exhaustive test for my issue.

What I was experiencing (basically that on subsequent deploys, HEAD was not being fetched from the remote), is no longer occurring against mainline devel.  So either it was an artifact of my pull, or otherwise resolved in devel, so I think case closed on this one.  Would like to get this test added to the PR though if possible, and drop my changes in #19257

robinro commented:
It seems all issues are fixed now and this can be merged.
For the test another PR is probably best.

sbussetti commented:
@robinro Done.  Cheers!

abadger commented:
Merged to devel and backported to stable-2.2.

Meabed commented:
I still have the error : 
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: UnboundLocalError: local variable 'remote_head' referenced before assignment
fatal: [10.100.90.54]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_gRy6vs/ansible_module_git.py\", line 1040, in <module>\n    main()\n  File \"/tmp/ansible_gRy6vs/ansible_module_git.py\", line 994, in main\n    result.update(changed=True, after=remote_head, msg='Local modifications exist')\nUnboundLocalError: local variable 'remote_head' referenced before assignment\n", "module_stdout": "", "msg": "MODULE FAILURE"}

robinro commented:
@Meabed Which version of the git module are you using? Did you clone the devel branch within the last 7 days? The error message is exactly as for the old version. So far the fix is not part of any public release.
